gpflow/features.py

```diff
@@ -0,0 +1,184 @@
+    def Kuu(self, kern, jitter=0.0):
+        """
+        Calculates the covariance matrix between features for kernel `kern`.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def Kuf(self, kern, Xnew):
+        """
+        Calculates the covariance matrix with function values at new points
+        `Xnew` for kernel `kern`.
+        """
+        raise NotImplementedError()
+
+
+class InducingPoints(InducingFeature):
+    """
+    Real-space inducing points
+    """
+
+    def __init__(self, Z):
+        """
+        :param Z: the initial positions of the inducing points, size M x D
+        """
+        super().__init__()
+        self.Z = Parameter(Z)
+
+    def __len__(self):
+        return self.Z.shape[0]
+
+    @decors.params_as_tensors
+    def Kuu(self, kern, jitter=0.0):
+        Kzz = kern.K(self.Z)
+        Kzz += jitter * tf.eye(len(self), dtype=settings.dtypes.float_type)
+        return Kzz
+
+    @decors.params_as_tensors
+    def Kuf(self, kern, Xnew):
+        Kzx = kern.K(self.Z, Xnew)
+        return Kzx
+
+
+class Multiscale(InducingPoints):
+    """
+    Multi-scale inducing features
+    Originally proposed in
+
+    ::
+
+      @incollection{NIPS2009_3876,
+        title = {Inter-domain Gaussian Processes for Sparse Inference using Inducing Features},
+        author = {Miguel L\'{a}zaro-Gredilla and An\'{\i}bal Figueiras-Vidal},
+        booktitle = {Advances in Neural Information Processing Systems 22},
+        year = {2009},
+      }
+
+    """
+
+    def __init__(self, Z, scales):
+        super().__init__(Z)
+        self.scales = Parameter(scales,
+                                transform=transforms.positive)  # Multi-scale feature widths (std. dev. of Gaussian)
+        assert self.Z.shape == scales.shape
+
+    def _cust_square_dist(self, A, B, sc):
+        """
+        Custom version of _square_dist that allows sc to provide per-datapoint length
+        scales. sc: N x M x D.
+        """
+        return tf.reduce_sum(tf.square((tf.expand_dims(A, 1) - tf.expand_dims(B, 0)) / sc), 2)
+
+    @decors.params_as_tensors
+    def Kuf(self, kern, Xnew):
+        if isinstance(kern, kernels.RBF):
+            with decors.params_as_tensors_for(kern):
+                Xnew, _ = kern._slice(Xnew, None)
+                Zmu, Zlen = kern._slice(self.Z, self.scales)
+                idlengthscales = kern.lengthscales + Zlen
+                d = self._cust_square_dist(Xnew, Zmu, idlengthscales)
+                Kuf = tf.transpose(kern.variance * tf.exp(-d / 2) *
+                                   tf.reshape(tf.reduce_prod(kern.lengthscales / idlengthscales, 1),
+                                              (1, -1)))
+            return Kuf
+        else:
+            raise NotImplementedError(
+                "Multiscale features not implemented for `%s`." % str(type(kern)))
+
+    @decors.params_as_tensors
+    def Kuu(self, kern, jitter=0.0):
+        if isinstance(kern, kernels.RBF):
```

Comment:
```
if not isinstance(kern, kernels.RBF):
    raise NotImplementedError(...)
....
```

Reply:
or use same approach with conditionals?

Reply:
Actually, we are thinking of doing a multiple dispatch in a different commit. It won't change the interface though.